### PR TITLE
highlight with treesitter for golang

### PR DESCRIPTION
**Pull request by tjdevries**
not too bad to start w/

### CODE REVIEW

**Comment by slimsag** [comment]
Q: Why do this in Rust instead of Go or TypeScript? The syntax highlighter being written in Rust has been a huge source of friction (the #1 reason IMO that nobody has really fixed the major issues with it in the past ~4 years at Sourcegraph)

**Comment by slimsag** [comment]
Stated another way: I don't personally feel qualified to review this, and I am wondering if we have general buy-in from the codeintel team that we're OK with using more Rust in general?

(I am totally fine with this approach if so, I just want to make sure there is a long term team owner for all of this that is not me)

**Comment by tjdevries** [comment]
This is an experiment right now, so we haven't committed to anything yet.

I can send you the link to the google doc where we've been discussing this (I'd have to ask @olafurpg  if he's OK w/ it being public), but this is just exploratory to compare some other options.

Short reasons to go about things this way:
- Official support for `tree-sitter` library
- Official `tree-sitter-highlight` library.
- CGO deployment for Go is difficult
- Already have syntect running and deployed, so augmenting with a tree-sitter service as a first step towards incorporating tree-sitter seems like good incremental progress as well as opening us up for lots of good opportunities in the future (using things like `tree-sitter-graph`, for example)
- Since tree-sitter does not have the same long tail of language support, we would probably need to continue using syntect for a long time, so it's possible that this method allows us to seamlessly transition and maintain both of those better than we have been able to do in the past (due to time constraints and prioritization).

We weren't planning on you being an owner (or having to be involved at all) for this project. Of course if you were really interested, I'm sure we could figure something out.

We'll see what happens after I get a bit further with exploring this angle. Thanks for checking in though and making your concerns known! I appreciate it.

**Comment by tjdevries** [comment]
Current status:

image

**Comment by tjdevries** [comment]
Should really consider implement `inherits` so we can re-use code more ez-pz:

```lua
--- Gets the list of files used to make up a query
---
---@param lang The language
---@param query_name The name of the query to load
---@param is_included Internal parameter, most of the time left as `nil`
function M.get_query_files(lang, query_name, is_included)
  local query_path = string.format('queries/%s/%s.scm', lang, query_name)
  local lang_files = dedupe_files(a.nvim_get_runtime_file(query_path, true))

  if #lang_files == 0 then return {} end

  local base_langs = {}

  -- Now get the base languages by looking at the first line of every file
  -- The syntax is the following :
  -- ;+ inherits: ({language},)*{language}
  --
  -- {language} ::= {lang} | ({lang})
  local MODELINE_FORMAT = "^;+%s*inherits%s*:?%s*([a-z_,()]+)%s*$"

  for _, file in ipairs(lang_files) do
    local modeline = safe_read(file, '*l')

    if modeline then
      local langlist = modeline:match(MODELINE_FORMAT)

      if langlist then
        for _, incllang in ipairs(vim.split(langlist, ',', true)) do
          local is_optional = incllang:match("%(.*%)")

          if is_optional then
            if not is_included then
              table.insert(base_langs, incllang:sub(2, #incllang - 1))
            end
          else
            table.insert(base_langs, incllang)
          end
        end
      end
    end
  end

  local query_files = {}
  for _, base_lang in ipairs(base_langs) do
    local base_files = M.get_query_files(base_lang, query_name, true)
    vim.list_extend(query_files, base_files)
  end
  vim.list_extend(query_files, lang_files)

  return query_files
end
```

Should not be hard to do.

File: docker-images/syntax-highlighter/Cargo.toml [hunk 1]
```diff
@@ -6,6 +6,32 @@ edition = "2018"
 license = "MIT"
 
 [dependencies]
+# TODO: Switch over to other one, but causes build problem.
```
Comment (varungandhi-src) [inline docker-images/syntax-highlighter/Cargo.toml:9]: This should be addressed if you rebase on top my recent changes in https://github.com/sourcegraph/sourcegraph/pull/29417

**Comment by slimsag** [comment]
👍 makes sense to me! I'm happy to step aside here and see what you end up landing on :) let me know if I can be of help in other ways

**Comment by olafurpg** [comment]
@tjdevries This approach looks very promising! I updated my vote in RFC 555 to go this route with a more detailed explanation why in the document https://docs.google.com/document/d/1IsC8NydIDQ3e_lEBqB1c084-u6BJc1dw0KmS2N0s2Eo/edit#

Just to document a few of the things we discussed offline, some changes we may want to implement:

- add a new `GET /lsif` endpoint to the syntax-highlighter service that returns LSIF typed from https://github.com/sourcegraph/sourcegraph/pull/29457
- update the LSIF typed schema to encode highlighting rules
- update the GraphQL `highlight()` endpoint to return the LSIF typed metadata
- update the `BlobPage` React component to render HTML from LSIF typed

I'd be happy to pair on any of these.

**Comment by tjdevries** [comment]
Closing in favor of #30016

**Pull request closed without merging**